pkg/reconciler/apis/crdcleanup/crdcleanup_controller.go

```diff
@@ -0,0 +1,260 @@
+
+	for _, boundResource := range binding.Status.BoundResources {
+		key := kcpcache.ToClusterAwareKey(apibinding.ShadowWorkspaceName.String(), "", boundResource.Schema.UID)
+		logging.WithQueueKey(logger, key).V(2).Info("queueing CRD via APIBinding")
+		c.queue.Add(key)
+	}
+}
+
+// Start starts the controller, which stops when ctx.Done() is closed.
+func (c *controller) Start(ctx context.Context, numThreads int) {
+	defer runtime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	logger := logging.WithReconciler(klog.FromContext(ctx), ControllerName)
+	ctx = klog.NewContext(ctx, logger)
+	logger.Info("Starting controller")
+	defer logger.Info("Shutting down controller")
+
+	for i := 0; i < numThreads; i++ {
+		go wait.UntilWithContext(ctx, c.startWorker, time.Second)
+	}
+
+	<-ctx.Done()
+}
+
+func (c *controller) startWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *controller) processNextWorkItem(ctx context.Context) bool {
+	// Wait until there is a new item in the working queue
+	k, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	key := k.(string)
+
+	logger := logging.WithQueueKey(klog.FromContext(ctx), key)
+	ctx = klog.NewContext(ctx, logger)
+	logger.V(4).Info("processing key")
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer c.queue.Done(key)
+
+	if err := c.process(ctx, key); err != nil {
+		runtime.HandleError(fmt.Errorf("%q controller failed to sync %q, err: %w", ControllerName, key, err))
+		c.queue.AddRateLimited(key)
+		return true
+	}
+	c.queue.Forget(key)
+	return true
+}
+
+func (c *controller) process(ctx context.Context, key string) error {
+	clusterName, _, name, err := kcpcache.SplitMetaClusterNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	obj, err := c.getCRD(clusterName, name)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil // object deleted before we handled it
+		}
+		return err
+	}
+
+	logger := logging.WithObject(klog.FromContext(ctx), obj)
+	ctx = klog.NewContext(ctx, logger)
+
+	if !apihelpers.IsCRDConditionTrue(obj, apiextensionsv1.Established) {
+		logger.V(4).Info("CRD is not established", "conditions", fmt.Sprintf("%#v", obj.Status.Conditions))
+		return nil
+	}
+
+	result, err := c.getAPIBindingsByBoundResourceUID(obj.Name)
+	if err != nil {
+		return err
+	}
+
+	// Index is used to quickly return
+	if len(result) > 0 {
+		return nil
+	}
+
+	// The index can't solely be relied on to be certain this CRD can be deleted.
+	// All bindings need to be up-to-date to ensure we don't erroneously delete the CRD.
+	bindingList, err := c.listAPIBindings()
+	if err != nil {
+		return err
+	}
+
+	for _, binding := range bindingList {
+		if binding.Status.Phase != apisv1alpha1.APIBindingPhaseBound || !conditions.IsTrue(binding, apisv1alpha1.BindingUpToDate) {
+			logger.V(3).Info("APIBinding has not completed initialization.  Need to wait until all bindings have bound resources")
+			c.queue.Add(key)
```

Comment:
Returning an error here will re-queue with backoff, let's do that instead of manually re-inserting & returning `nil`, which is generally used to mean that processing is done and the key should be forgotten.